src/software/simulation/threaded_simulator.h

```diff
@@ -0,0 +1,99 @@
+#pragma once
+
+#include "software/simulation/simulator.h"
+#include "software/proto/messages_robocup_ssl_wrapper.pb.h"
+#include <thread>
+#include <atomic>
+
+/**
+ * A simulator that runs in a separate thread and publishes protobuf
+ * data asynchronously.
+ */
+class ThreadedSimulator {
+public:
+    /**
+     * Creates a new ThreadedSimulator. The starting state of the simulation
+     * will have the given field, with no robots or ball.
+     *
+     * @param field The field to initialize the simulation with
+     */
+    explicit ThreadedSimulator(const Field& field);
+    ~ThreadedSimulator();
+
+    /**
+     * Registers the given callback function. This callback function will be
+     * called each time the simulation updates and a new SSL_WrapperPacket
+     * is generated.
+     *
+     * @param callback The callback function to register
+     */
+    void registerOnSSLWrapperPacketReadyCallback(const std::function<void(SSL_WrapperPacket)>& callback);
+
+    /**
+     * Starts running the simulator in a new thread. This is a non-blocking call.
+     * If the simulator is already running, this function does nothing.
+     */
+    void startSimulation();
+
+    /**
+     * Stops the simulation if it is running. If the simulation is not already
+     * running, this function does nothing.
+     */
+    void stopSimulation();
+
+    /**
+     * Sets the state of the ball in the simulation. No more than 1 ball may exist
+     * in the simulation at a time. If a ball does not already exist, a ball
+     * is added with the given state. If a ball already exists, it's state is set to the
+     * given state.
+     *
+     * @param ball_state The new ball state
+     */
+    void setBallState(const BallState& ball_state);
```

Comment:
We should explicitly note in each jdoc which of these functions are threadsafe.

Reply:
I (now) made all of them properly threadsafe, and indicated in all jdoc comments.